### ISSUE TITLE
Debug log level and https proxy environment variable

### ISSUE THREAD

**Issue by bbusioc** [InseeFrLab/legacy-onyxia-entrypoint#33]
Hello,

I'm currently trying to deploy Onyxia in our K8S cluster and I'm facing some issue, but I'm kind of stuck with the investigation.
I tend to believe that it might be related to the fact that from inside the pod, the ingresses that we set are not reachable unless we set a https proxy.
However, I can't see too much information in the pod logs, so I'm not sure that this is indeed what we are facing.
My questions would be:
1. Is there a way to increase the log verbosity for both ui and api compenents?
2. Is there a way to set http and https proxy environment variables for both ui and api components?

I'll try to sum up below the steps I'm performing, so maybe you notice anything wrong.
First I do the basic installation with helm:
helm repo add inseefrlab https://inseefrlab.github.io/helm-charts
helm install onyxia inseefrlab/onyxia --set ingress.enabled=true --set ingress.hosts[0].host=myhost.mydomain

When I try to access the ingress front-end, I get only a blank page, no errors in ui and api pods.
If I do an inspect on the blank page, I see these errors:
![image](https://user-images.githubusercontent.com/100717029/156341634-5f072e02-7555-437b-b655-4b564683b0ad.png)

Since it was saying something about keycloak, I tried adding quickly a keycloak in my namespace.
I linked Onyxia to my Keycloak installation using the below configs:

<pre>
ingress:
  enabled: true
  annotations:
    kubernetes.io/ingress.class: nginx
  hosts:
    - host: myhost.mydomain
ui:
  env:
    OIDC_REALM: onyxia-demo
    OIDC_CLIENT_ID: onyxia-client
    OIDC_URL: https://keycloak.myhost.mydomain/auth
api:
  env:
    keycloak.realm: onyxia-demo
    keycloak.auth-server-url: http://keycloak-http/auth
    authentication.mode: "openidconnect"
    springdoc.swagger-ui.oauth.clientId: onyxia-client
</pre>

After this the home page is visible, if I click on login I'm redirected to the keycloak login page, credentials are accepted and right after I get this pop-up error:
![image](https://user-images.githubusercontent.com/100717029/156342074-7db8eb02-b670-4264-9cc3-0ae95a6cbf8e.png)

I noticed that for onyxia-api there are some configs: http.proxyHost and http.proxyPort. (but nothing for https)
I tried using these ones and switched keycloak.auth-server-url from using a service name to the front-end of keycloak ingress; but I still have a connection timeout in onyxia-api pod logs.
Also tried https.proxyHost and https.proxyPort but it was the same. Not sure if these are taken into account anyway since they were not in the documentation.

So, without setting any http/https proxy and using service name for the keycloak url inside onyxia-api configuration, I still have the pop-up error mentioned above. ("There isn't an onyxia-ai hosted at...")
No errors in pod logs for both onyxia-ui and onyxia-api.

Any help on this would be great, mainly the 2 questions asked above because it will help me at least to better understand what's going on.

Thanks,
Bogdan Busioc

**Comment by olevitt** [comment]
Hi,

Thanks for your report.  
I think part of the issue you are facing is due to a recent change on the way onyxia bootstraps it's configuration from the API. The chart may be out of date, I will check this and get back to you.  

Regarding the log verbosity, there is currently no parameter to increase or decrease the verbosity.  
For proxy configuration :  
* The web module does not have any proxy configuration as it's a pure web application so all requests are done by your browser using your browser proxy configuration.  
* The API has a proxy configuration which is useful when the API pod does not have direct access to the internet. Onyxia API only requires internet access if you are using catalogs hosted on the internet (which is the case for the default catalog : https://github.com/inseefrlab/helm-charts-datascience) or if you intend to make it deploy onto a remote cluster (which is not the default configuration).  

I will check the state of the helm chart and get back to you.
Also, if you need more help for the installation / configuration, feel free to join this slack server : https://join.slack.com/t/3innovation/shared_invite/zt-12n4arfv4-Da3ywU3xeZn8mH6VMYkHdw

**Comment by bbusioc** [comment]
Hi,

Many thanks for your quick reply.
Actually in my case the https proxy would be useful not only for the default catalog, but also to reach keycloak.
In this basic/demo setup I'm using a keycloak instance started in the same namespace as onyxia api, so it's easy to use the service name; but eventually I will need to connect on the company open id connect solution which is hosted outside K8S and reachable only via a squid proxy.
But for this I think I might have a solution, so it's not such a big problem.
Anyway, I'll wait for your check on the helm chart state and I'll also join the slack server if any more questions will arise related to install/config.

Kind Regards,
Bogdan Busioc

**Comment by olevitt** [comment]
I narrowed down the issue to the API, created and merged this PR to fix it : https://github.com/InseeFrLab/onyxia-api/pull/120  
In a few minutes (typical CI / release build time) your onyxia should work again (once the API pod restarts).  

About your keycloak + proxy need : if your solution doesn't succeed there may be another approach : the Onyxia API only needs to contact keycloak once, at startup, to retrieve the public key from the realm. You can, if it fits your need, directly provide the public key from the realm to the onyxia API so that it doesn't need to contact keycloak at startup.

**Comment by bbusioc** [comment]
with no auth it seems to work, I'm getting now the home page and able to access the service catalog.
but with keycloak configured, I'm still getting this pop-up error after inserting my credentials:
![image](https://user-images.githubusercontent.com/100717029/156378778-353056b3-a88f-42bb-bd13-df3a5fbbab2b.png)

Any idea what might be causing this?

And regarding your solution to try and directly provide the public key for the realm, do you have any hints how to do this?
Is it this the config: keycloak.realm-public-key?

Thanks,
Bogdan

**Comment by garronej** [comment]
I have to update this error message, it is misleading... 
My guess is that your Keycloak config is set only on the front end and not on the backend.  
Currently it has to be on both.  

![image](https://user-images.githubusercontent.com/6702424/156394995-235e9469-44c4-4462-b884-69d20e513d12.png)

![image](https://user-images.githubusercontent.com/6702424/156395111-5b8f565d-13ec-4511-8159-e2c5fa19aa18.png)

**Comment by bbusioc** [comment]
It looks like an authentication problem indeed.
If I do an inspect on the page that gives the pop-up error I get this:
![image](https://user-images.githubusercontent.com/100717029/156406876-05c8eded-46ac-4392-b919-222681e658b5.png)

But I don't understand what I'm missing.
This is my config. If you can spot any error/missed configuration or have any other suggestions what I might check, please let me know.

<pre>
ingress:
  enabled: true
  annotations:
    kubernetes.io/ingress.class: nginx
  hosts:
    - host: onyxia.XXXXXXXXX.mydomain
ui:
  image:
    name: inseefrlab/onyxia-web
    version: 0.45.1
  env:
    OIDC_REALM: onyxia-demo
    OIDC_CLIENT_ID: onyxia-client
    OIDC_URL: https://keycloak.XXXXXXXXX.mydomain/auth
api:
  image:
    name: inseefrlab/onyxia-api
  env:
    security.cors.allowed_origins: "*"
    keycloak.resurce: onyxia-client
    keycloak.realm: onyxia-demo
    keycloak.auth-server-url: http://keycloak-http/auth
    keycloak.ssl-required: none
    keycloak.public-client: "true"
    keycloak.enable-basic-auth: "true"
    keycloak.bearer-only: "true"
    authentication.mode: "openidconnect"
    springdoc.swagger-ui.oauth.clientId: onyxia-client
    catalog.refresh.ms: "300000"
    http.proxyHost: squid.proxy.service.tech
    http.proxyPort: 3128
    http.noProxy: localhost,keycloak-http
</pre>

**Comment by olevitt** [comment]
Not sure if it's a copy / paste error or a real typo but you have `keycloak.resurce: onyxia-client` with a typo `resurce` instead of `resource`.  It should not matter much though.  
Please check your onyxia-api pod logs for any error. Onyxia-web is a web based application so error are read from the browser console. Onyxia-api is a server-side backend so use kubectl logs to retrieve logs.  
Last time I tried to use a non-https keycloak instance I failed. Not sure if it's possible or not. Anyway, you should have more info in the API logs.
The pop-up error you have is a catch-all error from the UI anytime the API responds a non-200 status code.  As Jo said, we will change it sometime in the future to something more informative. Note that you can use network debug from the browser dev tools to get a hopefully more interesting stacktrace in the reply from the API.
Yes, you can (theorically at least, havn't tried it myself), use `keycloak.realm-public-key` to provide the realm public key. If you do this then the Keycloak library from Onyxia-api should not try to contact keycloak.  See https://www.keycloak.org/docs/latest/securing_apps/#_java_adapter_config for more details

**Comment by bbusioc** [comment]
I don't know...
It doesn't seem to be from the connection between onyxia-api and keycloak, because I have this line in my logs:
<pre>
2022-03-03 07:52:09.370  INFO 1 --- [nio-8080-exec-2] o.keycloak.adapters.KeycloakDeployment   : Loaded URLs from http://keycloak:8080/auth/realms/onyxia-demo/.well-known/openid-configuration
</pre>

So, as far as I can understand the api connected successfully to keycloak.
There is nothing more in the onyxia-api pod logs, so it's quite hard to understand what's going on.

With network debug enabled on my browser I can see that I receive a 401 http code response on https://onyxia-frontend/api/user/info

Maybe I've missed something when I added the realm, client and user in my keycloak, but I can't figure out what it might be

**Comment by bbusioc** [comment]
I found the issue.
It was due to the "keycloak.auth-server-url" config, which I set it to use the K8S service name.
Because of this onyxia-api is expecting the issuer of the jwt token to be the http link with the K8S service name, but in fact it's the front-end URL from the ingress.
I tried to use the https front end URL for this setting, but I realized that the http proxy settings configured for accessing the service catalog was not used in this case.
So, I've set my proxy also with:
keycloak.proxy-url
Since the certificate of my front end is signed by an internal entity, I had to set also:
keycloak.disableTrustManager: "true"
keycloak.allowAnyHostname: "true"

Many settings and debugging hours, but in the end it works :)
Many thanks for the suggestion with developer tools and the link for the java adapter config.

From my point of view this issue may be closed.
